### PR TITLE
Populate external labels for annotations expansion 

### PR DESCRIPTION
**Pull request by sylr**
Attempt to fix https://github.com/prometheus/prometheus/issues/3043

### CODE REVIEW

**Comment by brian-brazil** [comment]
As noted in #3043, this is by design. In addition it'd be weird if external_labels were applied twice, and also weird if they ended up in ALERTS.

**Comment by sylr** [comment]
> external_labels were applied twice

@brian-brazil what do you mean ?

**Comment by sylr** [comment]
@fabxc @simonpasquier could I have your thoughts on this ?

**Comment by poblahblahblah** [comment]
We would **love** to see this merged in.

**Comment by beorn7** [comment]
I think we should do this, and I suspect the vast majority of prometheus-team thinks so, too. We should get clarity on this first, though, before actually implementing it. I have re-opened #3043 for now.

**Comment by beorn7** [comment]
I think this implementation is doing the right thing. Would be good to have a test for it, though (in particular about the case where a label already exists, which must then _not_ be replaced by the external label).

It might be good to use the same code path for adding the external labels to templates and to the externally served metrics.

**Comment by beorn7** [comment]
And yet another thought: If there are really concerns about the breakage this change might cause, we could hide this feature behind a flag (and then switch the default or remove it for Prom 3).

**Comment by sylr** [comment]
@beorn7 Would you prefer I unexport `*CurrentConfig` and add `GetCurrentConfig() *config` method ?

**Comment by beorn7** [comment]
@sylr Let's discuss implementation details once we have a decision on #3043 .

**Comment by beorn7** [comment]
I guess we have a decision in #3043 . @sylr Do you want to adjust this PR? I could totally understand if that long discussion and all the implications (like unit tests and console templates) is a bit overwhelming. In which case somebody else can take care of it...

**Comment by sylr** [comment]
I've been overwhelmed at work recently so I did not follow through the discussion but from I've been reading the concensus is to have a .ExternalLabels property ?

If I find the time I'll try to give it a look.

**Comment by beorn7** [comment]
So the plan is:
- Make the external-labels map from the config available in templates (alert annotations, alert labels, and console templates) as `.ExternalLabels`. (Pay attention that the template expansion also works on the alert status page.)
- To make it work within ranges, add a convenience variable `$external_labels` for alert templates.
- Add a way to feed external labels to rule unit testing (with `promtool`). Either by adding them to the test definition, or by passing in a Prometheus config file via the promtool command line, or both. (To be discussed.)

If you find time to work on this, it would be much appreciated.

**Comment by brian-brazil** [comment]
I wouldn't consider the last item in scope for this PR.

In relation to $ that would lead to an inconsistency where it's present in
alerts but not console templates. Or where that is in console templates,
but none of the others are. I believe all or nothing will resolve it, and
note that there's not much call for using this in a range.

**Comment by poblahblahblah** [comment]
Is there anything I can do to help push this change along?

**Comment by beorn7** [comment]
@sylr do you think you'll work on this in the foreseeable future?

Otherwise, this is just up for grabs for whoever feels qualified to do it.

**Comment by beorn7** [comment]
First step is really only introducing `.ExternalLabels`. The other points I listed don't have to be in this PR.

**Comment by sylr** [comment]
I'll take a look this afternoon.

**Comment by simonpasquier** [comment]
@sylr you can rebase on `master` to pass the CI.

**Comment by sylr** [comment]
@beorn7 PTAL

File: rules/alerting.go [hunk 1]
```diff
@@ -299,15 +301,25 @@ func (r *AlertingRule) Eval(ctx context.Context, ts time.Time, query QueryFunc,
...
+
+		// Labels.
 		for _, lbl := range smpl.Metric {
 			l[lbl.Name] = lbl.Value
 		}
 
-		tmplData := template.AlertTemplateData(l, smpl.V)
+		// External labels.
+		el := make(map[string]string)
+		if config.CurrentConfig != nil && (*config.CurrentConfig).GlobalConfig.ExternalLabels != nil {
```
Comment (brian-brazil) [inline rules/alerting.go]: There's a data race here

File: pkg/rulefmt/rulefmt.go [hunk 2]
```diff
@@ -155,8 +155,8 @@ func testTemplateParsing(rl *Rule) (errs []error) {
 	}
 
 	// Trying to parse templates.
-	tmplData := template.AlertTemplateData(make(map[string]string), 0)
-	defs := "{{$labels := .Labels}}{{$value := .Value}}"
+	tmplData := template.AlertTemplateData(make(map[string]string), make(map[string]string), 0)
+	defs := "{{$labels := .Labels}}{{$external_labels := .ExternalLabels}}{{$value := .Value}}"
```
Comment (brian-brazil) [inline pkg/rulefmt/rulefmt.go]: It feels odd to have Go style names for fields, and Javascript style names for variables.

Comment (sylr) [reply pkg/rulefmt/rulefmt.go]: My first choice was `$externalLabels` but I felt it looked bad.

File: rules/alerting.go (hunk 1 above)
Comment (sylr) [reply rules/alerting.go]: If GlobalConfig is nil ?

**Comment by sylr** [comment]
> 
> 
> This doesn't cover console templates.

No it does not, to be honnest I didn't even know about it.

File: rules/alerting.go (hunk 1 above)
Comment (sylr) [reply rules/alerting.go]: Should be fixed now.

File: pkg/rulefmt/rulefmt.go (hunk 2 above)
Comment (sylr) [reply pkg/rulefmt/rulefmt.go]: What do you propose ?

Comment (brian-brazil) [reply pkg/rulefmt/rulefmt.go]: externalLabels would be my thought.

Comment (sylr) [reply pkg/rulefmt/rulefmt.go]: Done.

**Comment by beorn7** [comment]
I think the CI failure was due to not being able to download staticcheck. Re-running…

**Comment by beorn7** [comment]
:tada: Build is green.
@brian-brazil is there anything left to improve in this PR?

File: config/config.go [hunk 3]
```diff
@@ -66,6 +67,10 @@ func LoadFile(filename string) (*Config, error) {
 
 // The defaults applied before parsing the respective config sections.
 var (
+	// CurrentConfig is a pointer to the current configuration.
```
Comment (brian-brazil) [inline config/config.go]: A global isn't how we pass config around, see https://github.com/prometheus/prometheus/blob/master/cmd/prometheus/main.go#L401

**Comment by brian-brazil** [comment]
> I think the CI failure was due to not being able to download staticcheck. Re-running…

That happens a bit unfortunately.

File: config/config.go (hunk 3 above)
Comment (sylr) [reply config/config.go]: I know.

**Comment by sylr** [comment]
> I'm still not seeing console templates covered.

Can you give me pointers ?

**Comment by brian-brazil** [comment]
Look for `ExpandHTML`.

**Comment by poblahblahblah** [comment]
Looks like the build is green. Can this make it's way into 2.8?

**Comment by brian-brazil** [comment]
I don't see this making it into 2.8, the code is still not right.

**Comment by beorn7** [comment]
Sorry for not following this review in the necessary depth to figure out why this is getting stale.

@sylr are you aware what makes @brian-brazil think this is not quite there yet? Or alternatively, @brian-brazil , could you explain in more detail?

**Comment by sylr** [comment]
@beorn7 I believe @brian-brazil does not like the config being accessed through a global var.

**Comment by beorn7** [comment]
Do you think it's a reasonable concern? Do you know how to address it?
(Just trying to unblock you two here.)

**Comment by sylr** [comment]
I don't know, more experienced go developer should answer that. What I know is that it provided an easy enough solution to implement that did not require to touch the current configuration reloading system which seems complex to me.

**Comment by brian-brazil** [comment]
I'm expecting the current system to be used as it's our standard way of doing things, and is better for those reusing our code.

**Comment by sylr** [comment]
What could be done to make it prettier is to:

- unexport config.CurrentConfig and add the needed function to config to set and get the current config
- move the config.CurrentConfigMutex inside the Config struct and add Public method to Lock/Unlock.

**Comment by brian-brazil** [comment]
RuleManager is already hooked into config updated, extend what's there.

**Comment by sylr** [comment]
So reverting this: https://github.com/prometheus/prometheus/commit/bd9f7460eb6fac35a103036d8fc5cd63c42ddd63

cc @fabxc

**Comment by brian-brazil** [comment]
You don't have to revert that.

**Comment by poblahblahblah** [comment]
Hey ya'll.

Is there anything we can do to help push this feature over the line in time for Prometheus 2.9.0?

**Comment by beorn7** [comment]
I hoped @sylr and @brian-brazil could shepherd this into a merge on their own. I would dive in here and help, but I am very short on time right now. Next week hopefully will be better...

**Comment by beorn7** [comment]
I finally got some time to look into this. After I found myself writing a lot of code changes myself, I decided to create #5463 on top of this. Please have a look there.

**Comment by beorn7** [comment]
Superseded by #5463.

**Pull request closed without merging**